### PR TITLE
AWAT minimum viable product

### PR DESCRIPTION
**Pull request by gigawhitlocks**
#### Summary

This pull request is **no longer a work in progress** and provides a minimum viable product for the Automatic Workspace Archive Translator, or AWAT.

The Autohealer and any other functionality laid out in [the design document](https://docs.google.com/document/d/1A3ewxbC2I22mQjtLOIyet2-qv_yb1d8MTDPE1rqimR0/edit) will come later.

Also, this PR should look **very familiar** because I am borrowing large amounts of code from the Provisioner repo. If there's anything that y'all think I should be bringing in as a library from that repo instead of copy-pasting, please yell at me about it! The `store` package is almost completely duplicated here, but it would require changes upstream to use as a library in this repo. That might be preferred, but I wasn't sure.

#### Updates:

This PR is now mostly ready to go. The full translation process is implemented at this time, as well as the basic API layer for controlling and monitoring the translation and import processes.

Updates will almost certainly be necessary as testing proceeds and as integration with the Provisioner is completed however this PR should provide a good starting point for this repo and for adding additional polish as we iterate.

The only thing definitely missing here is unit tests and I'm going to be working on those while receiving feedback on the implementation. **The core implementation here is complete at this time.**

### CODE REVIEW

**Comment by gigawhitlocks** [comment]
I'll add the SREs when this is closer to finished and again, no pressure about the review just yet since this is still in progress, but y'all can consider this PR a request for comments

**Review by gabrieljackson — COMMENTED** [review]
I took a skim through this and it generally looks good so far.

I didn't see anything obvious that we could have used a package for instead of copy pasta, but I think we should try to work towards this in the future. I believe Genesis will also have a bunch of shared code so it would be good if we could modularize things a bit.

Anyway, looks good so far. Request a review from me again when you get further along.

**Comment by gigawhitlocks** [comment]
Thanks Gabe! The snowpocalypse slowed me down a bit but I'll let you know when there are some significant changes to re-review

**Comment by gigawhitlocks** [comment]
Commit 21ab675 has a couple of design choices in it that I didn't think through thoroughly and if you're reviewing this code before I push more commits, after this comment, be aware that the changes in that commit are still more WIP than the stuff that came before it. 

Honestly I'm starting on a new logical step, beginning at 21ab675, and I'd be happy to peel off a spin-off branch & make a new PR if any of you feel that's necessary, but unless you do, I am working on the same branch in the meantime just in the interest of keeping my velocity up

File: internal/model/client.go
Comment (Szymongib) [inline internal/model/client.go:1]: Not sure if this is the case, but if we would like to use client and model in different services we probably want to move it out of `internal` directory as packages from it cannot be imported outside of the project itself.

**Review by Szymongib — COMMENTED** [review]
Great work so far Ian!
I have one question regarding terminology, in PR description you mention:
> - The complete database implementation for storing Transactions
> - an API for controlling Transactions

Is `Transaction` == `Translation`? I believe design doc also uses `Transaction` so I was slightly confused 😅

File: internal/model/client.go
Comment (gigawhitlocks) [reply internal/model/client.go:1]: Good call!

**Comment by gigawhitlocks** [comment]
`Is Transaction == Translation? I believe design doc also uses Transaction so I was slightly confused sweat_smile`

Yes, @Szymongib , I made the decision to just use the name Translation everywhere after a few commits at the beginning of the PR, because I found it less confusing in the code, but I didn't go back and retroactively update the design. Sorry for the confusion! Translation == Transaction.

**Comment by Szymongib** [comment]
Thanks for clarifying @gigawhitlocks! Good call on that, `Translation` seems like a better name.

File: internal/model/client.go
Comment (gigawhitlocks) [reply internal/model/client.go]: Done :)

**Comment by gigawhitlocks** [comment]
At this time, this PR is mostly stable again; I'm just adding a few more things to the command line tool for convenience during testing, development, and for SRE use, and that and some tests will be it for this PR. Once I have some tests and these final CLI commands added I'll remove the WIP status from this PR and begin working on integrating the AWAT with the Provisioner while addressing any comments.

Thanks!

**Comment by gigawhitlocks** [comment]
>  I'm just adding a few more things to the command line tool for convenience during testing, development, and for SRE use

this part is complete now

**Comment by gigawhitlocks** [comment]
@gabrieljackson you mentioned that you wanted to see this merged soon even if there's some polish that can be added after the merge. If you think this is ready to go (enough) could you click the approval button so there's a record that I didn't just go rogue and merge it unilaterally? :joy:

File: cmd/awat/import.go
Comment (gabrieljackson) [inline cmd/awat/import.go:17]: nit:
```suggestion
	getImportCmd.PersistentFlags().String(installationID, "", "ID of the installation associated with an import")
```

File: internal/api/import.go
Comment (gabrieljackson) [inline internal/api/import.go:52]: 0/5, but just returning the empty imports list is good as well. That's how the provisioner behaves if you would like to keep behavior similar.

File: internal/store/migrations.go
Comment (gabrieljackson) [inline internal/store/migrations.go:61]: nit: slight indentation issue

**Review by gabrieljackson — APPROVED** [review]
Awesome work! This is a great starting point that we can build off of. Only noted a few nits, but no blockers.

File: internal/slack/files.go
Comment (Szymongib) [inline internal/slack/files.go:126]: nit: I would consider extracting logic executed inside loop to separate function.

File: internal/store/migrations.go
Comment (gigawhitlocks) [reply internal/store/migrations.go:61]: feck

File: internal/slack/translator.go
Comment (Szymongib) [inline internal/slack/translator.go:140]: I wonder if it would not be better to defer archive removal at the top level?

File: internal/supervisor/supervisor.go
Comment (Szymongib) [inline internal/supervisor/supervisor.go:66]: Not sure if it is intended but this will cause translation to be unlocked only after all `work` have been processed (after the loop finishes and function exits) which means this instance may at some point hold locks for all translations.

File: internal/supervisor/supervisor.go
Comment (Szymongib) [inline internal/supervisor/supervisor.go:58]: Correct me if I am wrong here but it seems to me that the following scenario is possible:
- We have 2 translations to work on and 2 instances running
- Instance A picks up translation 1 and starts working on it
- Instance B sees translation 1 as locked so it picks up translation 2
- Instance B finishes releasing lock.
- Instance A finishes and picks up translation 2 (again).

File: model/client.go
Comment (Szymongib) [inline model/client.go:30]: I believe we should check for status codes in each clients method. From `http.Client` doc:
> A non-2xx status code doesn't cause an error.

As a safety net I guess it would also be good to defer closing of a response body.

File: model/translation_request.go
Comment (Szymongib) [inline model/translation_request.go:29]: 👍  😄

**Review by Szymongib — APPROVED** [review]
Excited for the Slack import!
I have spotted some more or less minor things but not sure which of them you planned to address in following PRs, so I leave it up to you.

File: go.mod
Comment (gigawhitlocks) [reply go.mod:7]: Yeah should be able to remove this now; will do, thanks for calling it out

File: internal/supervisor/supervisor.go
Comment (gigawhitlocks) [reply internal/supervisor/supervisor.go:66]: Nope, not intended, good eye. You're absolutely right.

File: model/client.go
Comment (gigawhitlocks) [reply model/client.go:30]: :+1: I was thinking about this a bit when looking at one of your recent PRs, where you used a bunch of `switch` statements for similar methods. I'll alter this to follow the same pattern.

File: internal/supervisor/supervisor.go
Comment (gigawhitlocks) [reply internal/supervisor/supervisor.go:58]: Ah, I considered the scenario where B is still working on translation 2 (in which case A would fail to lock it) but overlooked the scenario where 2 is completed. 

I am going to change this so that instead of getting a list of work, each time it queries it will get back one Translation instead of a list of them. Basically it will use, the same pattern used in `GetAndClaimNextReadyImport`. 

I'd already been considering that but thought I had considered all possible race conditions here, so didn't think it was necessary to change my initial implementation. Great catch @Szymongib

File: internal/slack/files.go
Comment (gigawhitlocks) [reply internal/slack/files.go:98]: :+1: this is actually copy-pasted from the mattermost-slack-exporter so I just didn't think about the signature of this helper function haha

Happy to change it though

File: internal/slack/files.go
Comment (gigawhitlocks) [reply internal/slack/files.go:60]: yeah, as mentioned in another comment this file is mostly stolen from the slack-advanced-exporter and so it was my intent to go through this and replace the `fmt` calls with log calls but I forgot about it, thanks

File: internal/slack/translator.go
Comment (gigawhitlocks) [reply internal/slack/translator.go:140]: Hmm, I know I specifically chose not to use a defer here but I can't remember why anymore :stuck_out_tongue: . I think I did this either to ease some pain related to development/testing or because I wanted to capture that error message? But I can't remember what pain I was alleviating and there's no reason I can't capture the error message in a deferred inline func so I'll do that..

File: internal/slack/files.go
Comment (gigawhitlocks) [reply internal/slack/files.go:1]: I agree, it's just proving to be a real pain to generate an appropriate testing artifact to process, which is what I'm working on / have been working on the last couple of days. The artifact I have been using for manual testing has some sensitive data in it, so I'm working on something to scrub the sensitive data from the archive so that it can be used in tests.

It's also possible that I might be able to refactor this file somewhat to make it more testable. It's borrowed from the slack-advanced-exporter which was, in my understanding, written in a bit of a hurry, so I just haven't taken the time to make it more easily testable or more modular.

File: build/Dockerfile
Comment (gigawhitlocks) [reply build/Dockerfile:4]: There was, yes, I wanted to use a utility that wasn't available in Alpine, but there is no longer any good reason because I went a different way with implementation. There are a couple of things that are wrong in this Dockerfile now, so I'll adjust before merging.

Comment (gigawhitlocks) [reply build/Dockerfile:4]: That said I wouldn't be surprised if we need to make some changes to this Dockerfile as we're preparing to deploy this service

File: internal/slack/files.go
Comment (gigawhitlocks) [reply internal/slack/files.go:98]: I actually really hate that shadowed variable and hadn't noticed it so thanks again for calling this out

File: internal/supervisor/supervisor.go
Comment (gigawhitlocks) [reply internal/supervisor/supervisor.go]: Fixing the race resolved below by returning 1 Translation instead of 10 also resolves this

**Pull request merged**
